src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/DepthFirstScanner.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.graphanalysis;
+
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.NotThreadSafe;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+
+/** Does a simple and somewhat efficient depth-first search of all FlowNodes in the DAG.
+ *
+ *  <p>Iteration order: depth-first search, revisiting parallel branches once done.
+ *  With parallel branches, the first branch is explored, then remaining branches are explored in reverse order.
+ *
+ * <p> The behavior is analogous to {@link org.jenkinsci.plugins.workflow.graph.FlowGraphWalker} but faster.
+ *  @author Sam Van Oort
+ */
+@NotThreadSafe
+public class DepthFirstScanner extends AbstractFlowScanner {
+
+    protected ArrayDeque<FlowNode> queue;
+
+    protected HashSet<FlowNode> visited = new HashSet<FlowNode>();
+
+    protected void reset() {
+        if (this.queue == null) {
+            this.queue = new ArrayDeque<FlowNode>();
+        } else {
+            this.queue.clear();
+        }
+        this.visited.clear();
+        this.myCurrent = null;
+    }
+
+    @Override
+    protected void setHeads(@Nonnull Collection<FlowNode> heads) {
```

Comment:
🐛 <code>myNext</code> is not reset in reset(). If setHeads() receives empty heads, myNext will be pointing to something from the previous heads list. And hasNext() will be returning true in such case